### PR TITLE
Special task force unicorn

### PR DESCRIPTION
**Pull request by fhennig**
## Review Checklist
- [x] Code contains useful comments
- [x] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
- [x] Double check docker usage

### CODE REVIEW

**Comment by fhennig** [comment]
Oh, I forgot some clippy warnings. Also I'll do the merging from main

**Comment by fhennig** [comment]
And here's the PR for the integration tests: https://github.com/stackabletech/integration-tests/pull/59

File: rust/operator-binary/Cargo.toml [hunk 1]
```diff
@@ -1,40 +1,36 @@
...
 authors = ["Lars Francke <lars.francke@stackable.de>"]
 description = "Stackable Operator for druid"
-edition = "2018"
+edition = "2021"
 license = "OSL-3.0"
-name = "stackable-druid-operator-binary"
+name = "stackable-druid-operator"
 repository = "https://github.com/stackabletech/druid-operator"
 version = "0.2.0-nightly"
 build = "build.rs"
```
Comment (siegfriedweber) [inline rust/operator-binary/Cargo.toml:9]: The `build` configuration seems not to be necessary anymore.

File: rust/operator-binary/Cargo.toml [hunk 2]
```diff
@@ -1,40 +1,36 @@
...
 
 [dependencies]
```
Comment (siegfriedweber) [inline rust/operator-binary/Cargo.toml:11]: In my opinion, the patch version of the dependencies is not useful in the `Cargo.toml`.

There are also some outdated versions listed.

File: rust/operator-binary/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+mod built_info {
+    include!(concat!(env!("OUT_DIR"), "/built.rs"));
+}
+
+pub const APP_NAME: &str = "druid";
+pub const APP_PORT: u16 = 2181;
+
+#[derive(StructOpt)]
+#[structopt(about = built_info::PKG_DESCRIPTION, author = "Stackable GmbH - info@stackable.de")]
```
Comment (siegfriedweber) [inline rust/operator-binary/src/main.rs:35]:
```suggestion
#[structopt(about = built_info::PKG_DESCRIPTION, author = stackable_operator::cli::AUTHOR)]
```

File: rust/crd/Cargo.toml [hunk 4]
```diff
@@ -1,15 +1,14 @@
 [package]
 authors = ["Lars Francke <lars.francke@stackable.de>"]
```
Comment (siegfriedweber) [inline rust/crd/Cargo.toml:2]:
```suggestion
authors = ["Stackable GmbH <info@stackable.de>"]
```

File: rust/operator-binary/Cargo.toml [hunk 5]
```diff
@@ -1,40 +1,36 @@
 [package]
 authors = ["Lars Francke <lars.francke@stackable.de>"]
```
Comment (siegfriedweber) [inline rust/operator-binary/Cargo.toml:2]:
```suggestion
authors = ["Stackable GmbH <info@stackable.de>"]
```

File: rust/operator-binary/src/druid_controller.rs [hunk 6]
```diff
@@ -0,0 +1,632 @@
...
+        "docker.stackable.tech/stackable/druid:{}-stackable{}",
+        version.to_string(),
+        DEFAULT_IMAGE_VERSION
+    )
+}
+
+pub fn druid_version(druid: &DruidCluster) -> Result<&str> {
+    Ok(&druid.spec.version)
+    //(Ok("0.22.0")
+    //Ok(&druid.spec.version.to_string()) // TODO
```
Comment (siegfriedweber) [inline rust/operator-binary/src/druid_controller.rs:625]: Fix or remove

File: rust/operator-binary/src/druid_controller.rs [hunk 7]
```diff
@@ -0,0 +1,632 @@
...
+    #[snafu(display("object {} has no namespace", obj_ref))]
+    ObjectHasNoNamespace { obj_ref: ObjectRef<DruidCluster> },
+    #[snafu(display("object {} defines no version", obj_ref))]
+    ObjectHasNoVersion { obj_ref: ObjectRef<DruidCluster> },
+    #[snafu(display("{} has no server role", obj_ref))]
+    NoServerRole { obj_ref: ObjectRef<DruidCluster> },
+    #[snafu(display("failed to calculate global service name for {}", obj_ref))]
+    GlobalServiceNameNotFound { obj_ref: ObjectRef<DruidCluster> },
+    #[snafu(display("failed to calculate service name for role {}", rolegroup))]
+    RoleGroupServiceNameNotFound { rolegroup: RoleGroupRef },
```
Comment (siegfriedweber) [inline rust/operator-binary/src/druid_controller.rs:65]: not used

File: rust/operator-binary/src/druid_controller.rs [hunk 8]
```diff
@@ -0,0 +1,632 @@
...
+    #[snafu(display("invalid product config for {}", druid))]
+    InvalidProductConfig {
+        source: stackable_operator::error::Error,
+        druid: ObjectRef<DruidCluster>,
+    },
+    #[snafu(display("failed to serialize zoo.cfg for {}", rolegroup))]
+    SerializeZooCfg {
+        source: stackable_operator::product_config::writer::PropertiesWriterError,
+        rolegroup: RoleGroupRef,
+    },
```
Comment (siegfriedweber) [inline rust/operator-binary/src/druid_controller.rs:100]: not used

File: rust/operator-binary/src/druid_controller.rs [hunk 9]
```diff
@@ -0,0 +1,632 @@
...
+        .await
+        .with_context(|| GetZookeeperConnStringConfigMap {
+            cm_name: zk_confmap.clone(), // TODO
+            namespace: zk_cm_ns.clone(), // TODO
+        })?
+        .data
+        .and_then(|mut data| data.remove("ZOOKEEPER"))
+        .with_context(|| MissingZookeeperConnString {
+            cm_name: zk_confmap.clone(), // TODO
+            namespace: zk_cm_ns.clone(), // TODO
```
Comment (siegfriedweber) [inline rust/operator-binary/src/druid_controller.rs:148]: What is the purpose of all these `TODOs`?

File: rust/operator-binary/src/druid_controller.rs [hunk 10]
```diff
@@ -0,0 +1,632 @@
...
+
+    roles.insert(
+        DruidRole::MiddleManager.to_string(),
+        (config_files.clone(), druid.spec.middle_managers.clone()),
+    );
+
+    roles.insert(
+        DruidRole::Router.to_string(),
+        (config_files, druid.spec.routers.clone()),
+    );
```
Comment (siegfriedweber) [inline rust/operator-binary/src/druid_controller.rs:183]: Copy&Paste where `DruidCluster::get_role` could be used.

File: rust/operator-binary/src/druid_controller.rs [hunk 11]
```diff
@@ -0,0 +1,632 @@
...
+
+    let role_config = transform_all_roles_to_config(&druid, roles);
+    let validated_role_config = validate_all_roles_and_groups_config(
+        &druid.spec.version.to_string(),
```
Comment (siegfriedweber) [inline rust/operator-binary/src/druid_controller.rs:187]: Why not use `druid_version(druid: &DruidCluster)` here?

File: rust/operator-binary/src/druid_controller.rs [hunk 12]
```diff
@@ -0,0 +1,632 @@
...
+                /*
+                                if let Some(zk_info) = &druid.spec.zookeeper_info {
+                                    transformed_config.insert(
+                                        ZOOKEEPER_CONNECTION_STRING.to_string(),
+                                        Some(zk_info.connection_string.clone()),
+                                    );
+                                } else {
+                                    return Err(error::Error::ZookeeperConnectionInformationError);
+                                }
+                */
```
Comment (siegfriedweber) [inline rust/operator-binary/src/druid_controller.rs:321]: Use or remove

**Review by siegfriedweber — CHANGES_REQUESTED** [review]
Additionaly to the comments in the code, I have some remarks:

* [ ] The reconciler is called every two seconds but it should only be called on changes.
* [x] Rebase onto the main branch because it contains changes necessary for this PR.
* [ ] Changelog is not updated.
* [ ] The installation with Helm is not described in the documentation.
* [x] The `packaging` directory still exists but is not used anymore.
* [x] `rust/operator/Cargo.toml` still exists but is not used anymore.
* [x] `rust/operator-binary/stackable-druid-operator.rs` still exists but is not used anymore.
* [x] Remove the section `features` in `rust/crd/Cargo.toml` if it is not used.
* [x] `rust/crd/src/error.rs` is not used and can be removed.
* [x] The following constants in `rust/crd/src/lib.rs` are not used and can probably be removed:
  * `CONF_DIR`
  * `DRUID_SERVICE`
  * `EXT_HDFS_STORAGE`
  * `DS_DIRECTORY`
  * `S3_ACCESS_KEY`
  * `S3_SECRET_KEY`

File: rust/operator-binary/Cargo.toml (hunk 1 above)
Comment (fhennig) [reply rust/operator-binary/Cargo.toml:9]: Oh yes indeed! build.rs is a special name that works without the config line there (https://doc.rust-lang.org/cargo/reference/build-scripts.html)

File: rust/operator-binary/Cargo.toml (hunk 2 above)
Comment (fhennig) [reply rust/operator-binary/Cargo.toml:11]: good point, will change to minor versions.

File: rust/operator-binary/src/druid_controller.rs (hunk 10 above)
Comment (fhennig) [reply rust/operator-binary/src/druid_controller.rs]: https://github.com/stackabletech/druid-operator/pull/59/commits/e60228e977488133a8983449fccfcfb1f3ce1fa6

**Comment by fhennig** [comment]
Oh, I missed the "The reconciler is called every two seconds but it should only be called on changes.". But that's actually something where I am unsure what to do. Could you maybe elaborate?

**Comment by fhennig** [comment]
The reconciler was watching endpoints by mistake. This has been removed now, thus the issue was fixed.

**Pull request merged**